### PR TITLE
Improve OpenMP exception handling

### PR DESCRIPTION
**Pull request by ldesreumaux**
See https://github.com/dmlc/xgboost/issues/6664.

All OpenMP regions now use an exception collector to rethrow the exceptions in the main thread.

### CODE REVIEW

**Comment by trivialfis** [comment]
I think dmlc-core has other utilities for omp like `dmlc::OMPException`.

File: R-package/src/xgboost_R.cc [hunk 1]
```diff
@@ -280,11 +296,15 @@ SEXP XGBoosterBoostOneIter_R(SEXP handle, SEXP dtrain, SEXP grad, SEXP hess) {
       << "gradient and hess must have same length";
   int len = length(grad);
   std::vector<float> tgrad(len), thess(len);
+  OMP_INIT();
```
Comment (trivialfis) [inline R-package/src/xgboost_R.cc:299]: Could you please consider using `dmlc::OMPException` instead of macros?

Comment (ldesreumaux) [reply R-package/src/xgboost_R.cc]: Done in 4fb1b75.

File: src/tree/updater_quantile_hist.cc [hunk 2]
```diff
@@ -825,14 +836,17 @@ void QuantileHistMaker::Builder<GradientSumT>::InitData(const GHistIndexMatrix&
...
-          for (size_t i = ibegin; i < iend; ++i) {
-           p_row_indices[i] = i;
-          }
+          exc.Run([&]() {
+            const size_t tid = omp_get_thread_num();
+            const size_t ibegin = tid * block_size;
+            const size_t iend = std::min(static_cast<size_t>(ibegin + block_size),
+                static_cast<size_t>(info.num_row_));
+            for (size_t i = ibegin; i < iend; ++i) {
+            p_row_indices[i] = i;
```
Comment (trivialfis) [inline src/tree/updater_quantile_hist.cc:845]: Indentation.

**Review by trivialfis — COMMENTED** [review]
Looks good to me overall.  Thanks for the huge refactor!  I will fetch the code later and try it out.

**Pull request merged**